[libvirt cluster] Use net-dhcp-leases to find VMs’ IPs

lhuard1A opened the pull request:
Query libvirt’s DHCP leases rather than inspecting the host’s ARP cache
to find the VMs’ IPs.

We used to ask libvirt for the MAC addresses of the VMs and then to inspect the host’s ARP cache to find the assigned IP.
There is in fact a way to query libvirt for the DHCP leases of the VMs.

lhuard1A commented:
It seems that Jenkins job failed because the libvirt python module isn’t installed on the Jenkins box. Is there anything I can do about that?

detiber commented:
@kwoodson can you coordinate getting the libvirt-python installed on the jenkins host?

@twiest :+1: from me on this change.

kwoodson commented:
ok to test

kwoodson commented:
@detiber @lhuard1A libvirt-python fixed.

lhuard1A commented:
Thanks @kwoodson , Jenkins now shows the following error:
```
E:143,34: Instance of 'virNetwork' has no 'DHCPLeases' member (no-member)
```
I don’t have that error on my PC.
Maybe this is coming from a different version of libvirt-python.
As @detiber said me that we shouldn’t integrate stuff depending on things not available of Fedora, I checked the Fedora version of libvirt-python and it seems that, on that version, DHCPLeases exists in the virNetwork class.

```
docker run -ti --rm fedora /bin/bash
yum install libvirt-python
egrep "class|def" /usr/lib64/python2.7/site-packages/libvirt.py | grep -A 4 virNetwork
class virNetwork(object):
    def __init__(self, conn, _obj=None):
    def __del__(self):
    def connect(self):
    def DHCPLeases(self, mac=None, flags=0):
```

Which OS is Jenkins running on?

detiber commented:
@lhuard1A It looks like this feature was added in libvirt 1.2.6.

I say we add a test in to verify that the libvirt package is >= 1.2.6 and fail otherwise.  I'm fine with limiting the libvirt support to RHEL7/Fedora (or other os's that have libvirt >= 1.2.6), we just need to let the user know.

kwoodson commented:
ok to test

kwoodson commented:
ok to test

twiest commented:
This all looks fine to me. @kwoodson when you get the bot stuff figured out (and it's passing), please merge this.

:+1:

kwoodson commented on inventory/libvirt/hosts/libvirt_generic.py:
```diff
@@ -133,18 +129,21 @@ def get_inventory(self):
+                if source_elem is not None and \
+                   mac_elem    is not None:
+                    dhcp_leases = conn.networkLookupByName(source_elem.get('network')) \
+                                      .DHCPLeases(mac_elem.get('address'))
```
Please add this above the DHCPLeases libvirt call:
```
# Adding this to disable pylint check specifically 
# ignoring libvirt-python versions that
# do not include DHCPLeases
# This is needed until we upgrade the build bot to 
# RHEL7 (>= 1.2.6 libvirt)
# pylint: disable=no-member
```

kwoodson commented:
@lhuard1A,

I have pushed a commit that updates the pylintrc file to exclude 'fixme' checks and also increased local variable counts.  The changes won't land in your commit until you merge those into your branch or after we merge this to master.

Please fix is the docstring messages, the previous pylint disable for no-member mentioned above, and the variable name length and we'll get this merged.
```
C: 49, 0: Missing class docstring (missing-docstring)
C: 66, 4: Missing method docstring (missing-docstring)
C: 73, 4: Missing method docstring (missing-docstring)
C: 96, 4: Missing method docstring (missing-docstring)
C:101, 4: Missing method docstring (missing-docstring)
C:129,12: Invalid variable name "ns" (invalid-name)
C:154, 0: Missing function docstring (missing-docstring)
C:160, 0: Missing function docstring (missing-docstring)
```

lhuard1A commented:
@kwoodson 
I rebased this PR in order to put it on top of your `pylintrc` update.
I fixed all the remaining pylint errors. (`pylint: disable=no-member`, the docstrings, and the `ns` variable name.
Jenkins is now happy! ;-)

twiest commented:
@lhuard1A thanks for this change! :)

:+1:
